packages/protocol/contracts/layer1/based2/AbstractInbox.sol

```diff
@@ -0,0 +1,350 @@
+    // -------------------------------------------------------------------------
+    // Constructor
+    // -------------------------------------------------------------------------
+
+    constructor() EssentialContract() { }
+
+    // -------------------------------------------------------------------------
+    // External Functions
+    // -------------------------------------------------------------------------
+
+    /// @notice Initializes the contract with owner and genesis block hash
+    /// @param _owner The owner address
+    /// @param _genesisBlockHash The genesis block hash
+    /// @param _gasIssuancePerSecond The initial gas issuance per second
+    function init4(
+        address _owner,
+        bytes32 _genesisBlockHash,
+        uint32 _gasIssuancePerSecond
+    )
+        external
+        initializer
+    {
+        _init(_owner, _genesisBlockHash, _gasIssuancePerSecond);
+    }
+
+    /// @inheritdoc IPropose
+    function propose4(
+        bytes calldata _packedSummary,
+        bytes calldata _packedBatches,
+        bytes calldata _packedEvidence,
+        bytes calldata _packedTrans
+    )
+        external
+        override(I, IPropose)
+        nonReentrant
+        returns (I.Summary memory)
+    {
+        I.Summary memory summary = _validateSummary(_packedSummary);
+        I.Batch[] memory batches = LibCodec.unpackBatches(_packedBatches);
+        I.BatchProposeMetadataEvidence memory evidence =
+            LibCodec.unpackBatchProposeMetadataEvidence(_packedEvidence);
+        I.TransitionMeta[] memory transitionMetas = LibCodec.unpackTransitionMetas(_packedTrans);
+
+        I.Config memory config = _getConfig();
+        LibState.Access memory access = _getReadWrite();
+
+        // Propose batches
+        summary = LibPropose.propose(config, access, summary, batches, evidence);
+
+        // Verify batches
+        summary = LibVerify.verify(config, access, summary, transitionMetas);
+
+        _saveSummaryHash(keccak256(abi.encode(summary)));
+        return summary;
+    }
+
+    /// @inheritdoc IProve
+    function prove4(
+        bytes calldata _packedSummary,
+        bytes calldata _packedBatchProveInputs,
+        bytes calldata _proof
+    )
+        external
+        override(I, IProve)
+        nonReentrant
+        returns (I.Summary memory)
+    {
+        I.Summary memory summary = _validateSummary(_packedSummary);
+        I.BatchProveInput[] memory inputs = LibCodec.unpackBatchProveInputs(_packedBatchProveInputs);
+
+        I.Config memory config = _getConfig();
+        LibState.Access memory access = _getReadWrite();
+
+        // Prove batches and get aggregated hash
+        bytes32 aggregatedBatchHash;
+        (summary, aggregatedBatchHash) = LibProve.prove(config, access, summary, inputs);
+
+        // Verify the proof
+        IVerifier2(config.verifier).verifyProof(aggregatedBatchHash, _proof);
+
+        _saveSummaryHash(keccak256(abi.encode(summary)));
```

Comment:
Seems this is unnecessary here and we should also always avoid updating `summary` in `prove4`, right? Otherwise, if two provers want to prove their batches in the same L1 block, all with correct proofs, the second one will revert.

Reply:
Very true. Fixed.